### PR TITLE
Account for whether artificial subclasses are skipped on type caching

### PR DESCRIPTION
**Pull request by augusto2112**
MetadataReader caches types only with the metadata address as the key. However a type lookup can be requested skipping artificial subclasses or not. This makes the cached results incorrect if two requests for the same type, but skipping subclasses on one and not on the other, are made. Fix this by adding a second dimension to the cache key.

rdar://101519300

### CODE REVIEW

**Comment by augusto2112** [comment]
@swift-ci smoke test

File: include/swift/Remote/MetadataReader.h [hunk 1]
```diff
@@ -193,8 +193,18 @@ class MetadataReader {
   /// amounts of data when we encounter corrupt values for sizes/counts.
   static const uint64_t MaxMetadataSize = 1048576; // 1MB
 
-  /// A cache of built types, keyed by the address of the type.
-  std::unordered_map<StoredPointer, BuiltType> TypeCache;
+  /// Define a has function for a std::pair<StoredPointer, bool>
+  struct TypeCacheKeyHash {
+    std::size_t operator()(const std::pair<StoredPointer, bool> &x) const {
+      return std::hash<StoredPointer>()(x.first) ^ std::hash<bool>()(x.second);
```
Comment (mikeash) [inline include/swift/Remote/MetadataReader.h:199]: There's a hash_combine function in `stdlib/include/llvm/ADT/Hashing.h` that will do a little better job of combining these two.

**Review by mikeash — APPROVED** [review]
I wonder if there's any point to making `skipArtificialSubclasses` optional. Seems like we should probably just always do it. But that can be a task for another day.

**Comment by augusto2112** [comment]
> I wonder if there's any point to making `skipArtificialSubclasses` optional. Seems like we should probably just always do it. But that can be a task for another day.

Actually in LLDB there is at least one place where we want to not skip artificial subclasses, since we iterate over a type's superclasses ourselves, and if metadata reader silently skipped over them our results would be off.

I do agree that this should've probably be defaulted to true instead of false though.

**Comment by augusto2112** [comment]
@swift-ci smoke test

**Comment by augusto2112** [comment]
@swift-ci smoke test

File: include/swift/Remote/MetadataReader.h [hunk 2]
```diff
@@ -193,8 +195,18 @@ class MetadataReader {
...
+  struct TypeCacheKeyHash {
+    std::size_t operator()(const std::pair<StoredPointer, bool> &x) const {
+      return llvm::hash_combine(x.first, x.second);
+    }
+  };
+
+  /// A cache of built types, keyed by the address of the type and whether the
+  /// request ignored articial superclasses or not.
+  std::unordered_map<std::pair<StoredPointer, bool>, BuiltType,
+                     TypeCacheKeyHash>
```
Comment (adrian-prantl) [inline include/swift/Remote/MetadataReader.h:208]: Just curious: If we are using the LLVM libraries anyway, why not use a DenseMap?

Comment (augusto2112) [reply include/swift/Remote/MetadataReader.h:208]: Looks like we can, I'll change this in a separate patch.

**Comment by adrian-prantl** [comment]
@augusto2112 You will probably also want to cherry-pick this to `release/5.10`

**Pull request merged**